### PR TITLE
Add UngroupDetectors to ExtractMask

### PR DESCRIPTION
**Pull request by rosswhitfield**
### Description of work
Ref: [3108: Add 'InstrumentDonor' parameter to ExtractMask](https://ornlrse.clm.ibmcloud.com/ccm/resource/itemName/com.ibm.team.workitem.WorkItem/3108)

#### Summary of work

Fixes #xxxx. 

### To test:

```python
ws = CreateSampleWorkspace(NumBanks=4,BankPixelWidth=2)
groups = CreateGroupingWorkspace(ws, GroupDetectorsBy="bank")
ws = GroupDetectors(ws, CopyGroupingFromWorkspace="groups")
MaskDetectors(ws,WorkspaceIndexList=[0,2])

wsMask, maskList = ExtractMask(ws)

wsMask_ungrouped, maskList_ungrouped = ExtractMask(ws, UngroupDetectors=True)

```

---

### Reviewer

Please comment on the points listed below ([full description](http://developer.mantidproject.org/ReviewingAPullRequest.html)).
**Your comments will be used as part of the gatekeeper process, so please comment clearly on what you have checked during your review.** If changes are made to the PR during the review process then your final comment will be the most important for gatekeepers. In this comment you should make it clear why any earlier review is still valid, or confirm that all requested changes have been addressed.

#### Code Review

- Is the code of an acceptable quality?
- Does the code conform to the [coding standards](http://developer.mantidproject.org/Standards/)?
- Are the unit tests small and test the class in isolation?
- If there is GUI work does it follow the [GUI standards](http://developer.mantidproject.org/Standards/GUIStandards.html)?
- If there are changes in the release notes then do they describe the changes appropriately?
- Do the release notes conform to the [release notes guide](https://developer.mantidproject.org/Standards/ReleaseNotesGuide.html)?

#### Functional Tests

- Do changes function as described? Add comments below that describe the tests performed?
- Do the changes handle unexpected situations, e.g. bad input?
- Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### Gatekeeper

If you need to request changes to a PR then please add a comment and set the review status to "Request changes". This will stop the PR from showing up in the list for other gatekeepers.

### CODE REVIEW

**Review by jmborr — COMMENTED** [review]
@rosswhitfield you wrote:

> I don't know if I fully understand the behavior we are after. So from what I can gather we need ExtractMask to be able to create a workspace that only has one detector per spectra, so if the InputWorkspace has multiple detectors per spectra we would break out those spectra into multiple. I think what would be better than a new input workspace "InstrumentDonor" we could just have a flag that changes this behavior.

After looking at the changes, your suggestion makes much more sense than passing a workspace with the same instrument as the input workspace. I'd just include a boolean argument (`UngroupDetectors`). Is there a use-case when `inputWS` and `donorWS` have different instrument geometries, yet we still want to use the detector ID's from donorWS?

**Review by jmborr — APPROVED** [review]
Manual test runs as expected. Changes look :ok_hand:

File: Framework/Algorithms/src/ExtractMask.cpp [hunk 1]
```diff
@@ -62,26 +63,21 @@ void ExtractMask::exec() {
...
+  maskWS->setTitle(inputWS->getTitle());
 
-  PARALLEL_FOR_IF(Kernel::threadSafe(*inputWS, *maskWS))
-  for (int64_t i = 0; i < nHist; ++i) {
-    PARALLEL_START_INTERRUPT_REGION
-    bool inputIsMasked(false);
-    if (spectrumInfo.hasDetectors(i)) {
-      // special workspaces can mysteriously have the mask bit set
-      inputIsMasked = (inputWSIsSpecial && inputMaskWS->isMaskedIndex(i)) || spectrumInfo.isMasked(i);
+  // set mask from from detectorList
```
Comment (ekapadi) [inline Framework/Algorithms/src/ExtractMask.cpp:74]: It looks like we went from an OMP-threaded for loop to a non-threaded for loop.  Was there no other way to do this?  (If I assume masking is _rare_, this is probably OK.  I can imagine the opposite case, however.)

File: docs/source/algorithms/ExtractMask-v1.rst [hunk 2]
```diff
@@ -21,6 +21,11 @@ link is preserved so that the instrument view functions correctly.
 A list of masked detector IDs is also output. Note this contains the detector IDs which
 are masked rather than the index or spectrum number.
 
+Ungroup Detectors
+-----------------
+
+If ``UngroupDetectors`` is set to True, the detector IDs are ungrouped. This means that if the input workspace has a grouped detector, the output workspace will have a single spectra for each detector.
```
Comment (ekapadi) [inline docs/source/algorithms/ExtractMask-v1.rst:27]: Suggest:  **"This means that if the input workspace is grouped with multiple detectors per spectrum, the output workspace will have a single spectrum per detector."**  (Notes:  spectrum [singular], spectra [plural], are we using the term "detector" [ colloqual ] here or "pixel"?)

File: docs/source/release/v6.12.0/Framework/Algorithms/New_features/38198.rst [hunk 3]
```diff
@@ -0,0 +1 @@
+- A new property ``UngroupDetectors`` was added to :ref:`ExtractMask <algm-ExtractMask>` allowing the output ``MaskWorkspace`` to expand the spectra to individal detectors.
```
Comment (ekapadi) [inline docs/source/release/v6.12.0/Framework/Algorithms/New_features/38198.rst:1]: Same kind of thing as previous comments: suggest: "allowing the output ``MaskWorkspace`` to represent the mask flags of the individual detectors."

**Review by ekapadi — COMMENTED** [review]
I built the branch, and ran through the suggested testing procedure.  Everything works as it should.  The code quality looks good, and there are new unit tests to cover the behavior of new argument.  I only had a few small comments (i.e. suggestions) about the wording of some of the documentation.  Also, please take a look again at the _main_ loop and see if there's a straightforward way to keep the OMP threading.  If there's not, that's OK, but at least we will have looked at it.  (I'm thinking about over 64 million pixels, in case there's a large masking fraction, that loop can take a _long_ time!)

**Pull request merged**